pkg/model/iam.go

```diff
@@ -51,133 +51,158 @@ const RolePolicyTemplate = `{
 }`
 
 func (b *IAMModelBuilder) Build(c *fi.ModelBuilderContext) error {
-	// Collect the roles in use
-	var roles []kops.InstanceGroupRole
+	// Collect managed Instance Group roles
+	managedRoles := make(map[kops.InstanceGroupRole]bool)
+
+	// Collect Instance Profile ARNs and their associated Instance Group roles
+	sharedProfileARNsToIGRole := make(map[string]kops.InstanceGroupRole)
 	for _, ig := range b.InstanceGroups {
-		found := false
-		for _, r := range roles {
-			if r == ig.Spec.Role {
-				found = true
+		if ig.Spec.IAM != nil && ig.Spec.IAM.Profile != nil {
+			specProfile := fi.StringValue(ig.Spec.IAM.Profile)
+			if matchingRole, ok := sharedProfileARNsToIGRole[specProfile]; ok {
+				if matchingRole != ig.Spec.Role {
+					return fmt.Errorf("Found IAM instance profile assigned to multiple Instance Group roles %v and %v: %v",
+						ig.Spec.Role, sharedProfileARNsToIGRole[specProfile], specProfile)
+				}
+			} else {
+				sharedProfileARNsToIGRole[specProfile] = ig.Spec.Role
 			}
+		} else {
+			managedRoles[ig.Spec.Role] = true
 		}
-		if !found {
-			roles = append(roles, ig.Spec.Role)
+	}
+
+	// Generate IAM tasks for each shared role
+	for profileARN, igRole := range sharedProfileARNsToIGRole {
+		iamName, err := findCustomAuthNameFromArn(profileARN)
+		if err != nil {
+			return fmt.Errorf("unable to parse instance profile name from arn %q: %v", profileARN, err)
 		}
+		err = b.buildIAMTasks(igRole, iamName, c, true)
 	}
 
-	// Generate IAM objects etc for each role
-	for _, role := range roles {
-		name := b.IAMName(role)
+	// Generate IAM tasks for each managed role
+	for igRole := range managedRoles {
+		iamName := b.IAMName(igRole)
+		err := b.buildIAMTasks(igRole, iamName, c, false)
+		if err != nil {
+			return err
+		}
+	}
 
-		var iamRole *awstasks.IAMRole
-		{
-			rolePolicy, err := b.buildAWSIAMRolePolicy()
-			if err != nil {
-				return err
-			}
+	return nil
+}
 
-			iamRole = &awstasks.IAMRole{
-				Name:      s(name),
-				Lifecycle: b.Lifecycle,
+func (b *IAMModelBuilder) buildIAMTasks(igRole kops.InstanceGroupRole, iamName string, c *fi.ModelBuilderContext, shared bool) error {
+	var iamRole *awstasks.IAMRole
+	{
```

Comment:
Can we insert another level of indentation here - the indentation change makes the diff hard to review.

e.g.

```
{ // To minimize diff for easier code review
  { // To minimize diff for easier code review
...
```